deps/rabbit/src/rabbit_msg_store.erl

```diff
@@ -2096,28 +2105,28 @@ delete_file(File, State = #gc_state { file_summary_ets = FileSummaryEts,
 
 scan_and_vacuum_message_file(File, #gc_state{ index_ets = IndexEts, dir = Dir }) ->
     %% Messages here will be end-of-file at start-of-list
-    {ok, Messages, _FileSize} =
-        scan_file_for_valid_messages(Dir, filenum_to_name(File)),
-    %% foldl will reverse so will end up with msgs in ascending offset order
-    lists:foldl(
-      fun ({MsgId, TotalSize, Offset}, Acc = {List, Size}) ->
-              case index_lookup(IndexEts, MsgId) of
-                  #msg_location { file = File, total_size = TotalSize,
-                                  offset = Offset, ref_count = 0 } = Entry ->
-                      index_delete_object(IndexEts, Entry),
-                      Acc;
-                  #msg_location { file = File, total_size = TotalSize,
-                                  offset = Offset } = Entry ->
-                      {[ Entry | List ], TotalSize + Size};
-                  %% Fan-out may remove the entry but also write a new
-                  %% entry in a different file when it needs to write
-                  %% a message and the existing reference is in a file
-                  %% that's about to be deleted. So we explicitly accept
-                  %% these cases and ignore this message.
-                  #msg_location { file = OtherFile, total_size = TotalSize }
-                          when File =/= OtherFile ->
-                      Acc;
-                  not_found ->
-                      Acc
-              end
-      end, {[], 0}, Messages).
+    Path = form_filename(Dir, filenum_to_name(File)),
+    {ok, Messages, _FileSize} = scan_file_for_valid_messages(Path,
+        fun ({MsgId, TotalSize, Offset}) ->
+            case index_lookup(IndexEts, MsgId) of
+                #msg_location { file = File, total_size = TotalSize,
+                                offset = Offset, ref_count = 0 } = Entry ->
+                    index_delete_object(IndexEts, Entry),
+                    invalid;
```

Comment:
isn't this a valid message from scanning perspective? So we don't want to scan byte-by-byte just because ref count is zero? Maybe need a 3rd status `valid_but_ignore` (I might be wrong)

Reply:
It is but since we are deleting it in this clause it becomes no longer valid. Maybe I can add a comment.

Reply:
what I meant is something like the bellow where we jump to `Offset + TotalSize` instead of `Offset + 1` but we don't add the `Entry` to the `Acc` in `scan_data/7`
```
                %% Confirmed to be a message by the provided fun
                %% but we don't want to keep it.
                valid_but_ingore ->
                    scan_data(Rest, Fd, Fun, Offset + TotalSize, FileSize,
                              MsgIdsFound#{MsgIdInt => true}, Acc);
```
(This is just an efficiency optimisation, not a correctness issue)

Reply:
Oh fair enough, yes that would be better. Maybe it'll even help with file delete performance. Will check on Monday.

Reply:
Done, and that solves the slow delete. Thanks for noticing!!

Reply:
great